File: C/mediapipe_api/framework/timestamp.cc
#include "mediapipe_api/framework/timestamp.h"

MpReturnCode mp_Timestamp__l(int64 timestamp, mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp(timestamp);
    return MpReturnCode::Success;
  } CATCH_ALL
}

void mp_Timestamp__delete(mediapipe::Timestamp* timestamp) {
  delete timestamp;
}

int64 mp_Timestamp__Value(mediapipe::Timestamp* timestamp) {
  return timestamp->Value();
}

double mp_Timestamp__Seconds(mediapipe::Timestamp* timestamp) {
  return timestamp->Seconds();
}

int64 mp_Timestamp__Microseconds(mediapipe::Timestamp* timestamp) {
  return timestamp->Microseconds();
}

MpReturnCode mp_Timestamp__DebugString(mediapipe::Timestamp* timestamp, const char** str_out) {
  TRY {
    *str_out = strcpy_to_heap(timestamp->DebugString());
    return MpReturnCode::Success;
  } CATCH_ALL
}

bool mp_Timestamp__IsSpecialValue(mediapipe::Timestamp* timestamp) {
  return timestamp->IsSpecialValue();
}

bool mp_Timestamp__IsRangeValue(mediapipe::Timestamp* timestamp) {
  return timestamp->IsRangeValue();
}

bool mp_Timestamp__IsAllowedInStream(mediapipe::Timestamp* timestamp) {
  return timestamp->IsAllowedInStream();
}

MpReturnCode mp_Timestamp__NextAllowedInStream(mediapipe::Timestamp* timestamp, mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp { timestamp->NextAllowedInStream() };
    return MpReturnCode::Success;
  } CATCH_ALL
}

MpReturnCode mp_Timestamp__PreviousAllowedInStream(mediapipe::Timestamp* timestamp, mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp { timestamp->PreviousAllowedInStream() };
    return MpReturnCode::Success;
  } CATCH_ALL
}

MpReturnCode mp_Timestamp_FromSeconds__d(double seconds, mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp { mediapipe::Timestamp::FromSeconds(seconds) };
    return MpReturnCode::Success;
  } CATCH_ALL
}

MpReturnCode mp_Timestamp_Unset(mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp { mediapipe::Timestamp::Unset() };
    return MpReturnCode::Success;
  } CATCH_ALL
}

MpReturnCode mp_Timestamp_Unstarted(mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp { mediapipe::Timestamp::Unstarted() };
    return MpReturnCode::Success;
  } CATCH_ALL
}

MpReturnCode mp_Timestamp_PreStream(mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp { mediapipe::Timestamp::PreStream() };
    return MpReturnCode::Success;
  } CATCH_ALL
}

MpReturnCode mp_Timestamp_Min(mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp { mediapipe::Timestamp::Min() };
    return MpReturnCode::Success;
  } CATCH_ALL
}

MpReturnCode mp_Timestamp_Max(mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp { mediapipe::Timestamp::Max() };
    return MpReturnCode::Success;
  } CATCH_ALL
}

MpReturnCode mp_Timestamp_PostStream(mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp { mediapipe::Timestamp::PostStream() };
    return MpReturnCode::Success;
  } CATCH_ALL
}

MpReturnCode mp_Timestamp_OneOverPostStream(mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp { mediapipe::Timestamp::OneOverPostStream() };
    return MpReturnCode::Success;
  } CATCH_ALL
}

MpReturnCode mp_Timestamp_Done(mediapipe::Timestamp** timestamp_out) {
  TRY {
    *timestamp_out = new mediapipe::Timestamp { mediapipe::Timestamp::Done() };
    return MpReturnCode::Success;
  } CATCH_ALL
}

File: Assets/MediaPipe/SDK/Scripts/Framework/Timestamp.cs
using System;

namespace Mediapipe {
  public class Timestamp : MpResourceHandle, IEquatable<Timestamp> {
    public Timestamp(IntPtr ptr) : base(ptr) {}

    public Timestamp(Int64 value) : base() {
      UnsafeNativeMethods.mp_Timestamp__l(value, out var ptr);
      this.ptr = ptr;
    }

    protected override void DeleteMpPtr() {
      UnsafeNativeMethods.mp_Timestamp__delete(ptr);
    }

    #region IEquatable<Timestamp>
    public bool Equals(Timestamp other) {
      if (other == null) { return false; }

      return Microseconds() == other.Microseconds();
    }

    public override bool Equals(Object obj) {
      Timestamp timestampObj = obj == null ? null : (obj as Timestamp);

      return timestampObj != null && Equals(timestampObj);
    }

    public static bool operator ==(Timestamp x, Timestamp y) {
      if (((object)x) == null || ((object)y) == null) {
        return Object.Equals(x, y);
      }

      return x.Equals(y);
    }

    public static bool operator !=(Timestamp x, Timestamp y) {
      if (((object)x) == null || ((object)y) == null) {
        return !Object.Equals(x, y);
      }

      return !(x.Equals(y));
    }

    public override int GetHashCode() {
      return this.Microseconds().GetHashCode();
    }
    #endregion

    public Int64 Value() {
      return SafeNativeMethods.mp_Timestamp__Value(mpPtr);
    }

    public double Seconds() {
      return SafeNativeMethods.mp_Timestamp__Seconds(mpPtr);
    }

    public Int64 Microseconds() {
      return SafeNativeMethods.mp_Timestamp__Microseconds(mpPtr);
    }

    public bool IsSpecialValue() {
      return SafeNativeMethods.mp_Timestamp__IsSpecialValue(mpPtr);
    }

    public bool IsRangeValue() {
      return SafeNativeMethods.mp_Timestamp__IsRangeValue(mpPtr);
    }

    public bool IsAllowedInStream() {
      return SafeNativeMethods.mp_Timestamp__IsAllowedInStream(mpPtr);
    }

    public string DebugString() {
      return MarshalStringFromNative(UnsafeNativeMethods.mp_Timestamp__DebugString);
    }

    public Timestamp NextAllowedInStream() {
      UnsafeNativeMethods.mp_Timestamp__NextAllowedInStream(mpPtr, out var nextPtr).Assert();

      GC.KeepAlive(this);
      return new Timestamp(nextPtr);
    }

    public Timestamp PreviousAllowedInStream() {
      UnsafeNativeMethods.mp_Timestamp__PreviousAllowedInStream(mpPtr, out var prevPtr).Assert();

      GC.KeepAlive(this);
      return new Timestamp(prevPtr);
    }

    public static Timestamp FromSeconds(double seconds) {
      UnsafeNativeMethods.mp_Timestamp_FromSeconds__d(seconds, out var ptr).Assert();

      return new Timestamp(ptr);
    }

    #region SpecialValues
    public static Timestamp Unset() {
      UnsafeNativeMethods.mp_Timestamp_Unset(out var ptr).Assert();

      return new Timestamp(ptr);
    }

    public static Timestamp Unstarted() {
      UnsafeNativeMethods.mp_Timestamp_Unstarted(out var ptr).Assert();

      return new Timestamp(ptr);
    }

    public static Timestamp PreStream() {
      UnsafeNativeMethods.mp_Timestamp_PreStream(out var ptr).Assert();

      return new Timestamp(ptr);
    }

    public static Timestamp Min() {
      UnsafeNativeMethods.mp_Timestamp_Min(out var ptr).Assert();

      return new Timestamp(ptr);
    }

    public static Timestamp Max() {
      UnsafeNativeMethods.mp_Timestamp_Max(out var ptr).Assert();

      return new Timestamp(ptr);
    }

    public static Timestamp PostStream() {
      UnsafeNativeMethods.mp_Timestamp_PostStream(out var ptr).Assert();

      return new Timestamp(ptr);
    }

    public static Timestamp OneOverPostStream() {
      UnsafeNativeMethods.mp_Timestamp_OneOverPostStream(out var ptr).Assert();

      return new Timestamp(ptr);
    }

    public static Timestamp Done() {
      UnsafeNativeMethods.mp_Timestamp_Done(out var ptr).Assert();

      return new Timestamp(ptr);
    }
    #endregion
  }
}
